File: leetcode-cn/0053.0_Maximum_Subarray.py
'''
T: O(N)
S: O(1)

执行用时：100 ms, 在所有 Python3 提交中击败了76.84% 的用户
内存消耗：25.2 MB, 在所有 Python3 提交中击败了69.19% 的用户
通过测试用例：209 / 209
'''
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        ans = nums[0]
        pre = 0
        for x in nums:
            if pre >= 0:
                pre += x
            else:
                pre = x
            ans = max(ans, pre)
        return ans


'''
执行用时：148 ms, 在所有 Python3 提交中击败了76.08% 的用户
内存消耗：25.5 MB, 在所有 Python3 提交中击败了93.95% 的用户
通过测试用例：209 / 209
'''
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        ans = nums[0]
        pre = 0
        for x in nums:
            pre += x
            ans = max(ans, pre)
            if pre < 0:
                pre = 0

        return ans 


'''
segment tree
分治思想

执行用时：1064 ms, 在所有 Python3 提交中击败了5.03% 的用户
内存消耗：26.1 MB, 在所有 Python3 提交中击败了24.23% 的用户
通过测试用例：209 / 209
'''
class Status:
    def __init__(self, lSum, rSum, mSum, iSum):
        self.lSum = lSum 
        self.rSum = rSum 
        self.mSum = mSum 
        self.iSum = iSum 

def getInfo(nums, l, r) -> Status:
    if l == r:
        return Status(nums[l], nums[l], nums[l], nums[l])
    m = (l + r) >> 1
    lSub = getInfo(nums, l, m)
    rSub = getInfo(nums, m + 1, r)
    return pushUp(lSub, rSub)

def pushUp(l: Status, r: Status) -> Status:
    iSum = l.iSum + r.iSum 
    lSum = max(l.lSum, l.iSum + r.lSum)
    rSum = max(r.rSum, r.iSum + l.rSum)
    mSum = max(l.mSum, r.mSum, l.rSum + r.lSum)
    return Status(lSum, rSum, mSum, iSum)

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        return getInfo(nums, 0, len(nums) - 1).mSum 


'''
segment tree
分治思想

执行用时：1052 ms, 在所有 Python3 提交中击败了5.03% 的用户
内存消耗：25.7 MB, 在所有 Python3 提交中击败了58.52% 的用户
通过测试用例：209 / 209
'''
class Status:
    def __init__(self, lSum, rSum, mSum, iSum):
        self.lSum = lSum 
        self.rSum = rSum 
        self.mSum = mSum 
        self.iSum = iSum 

    @classmethod
    def getInfo(cls, nums: List[int], l: int, r: int):
        if l == r:
            return Status(nums[l], nums[l], nums[l], nums[l])
        m = (l + r) >> 1
        lSub = cls.getInfo(nums, l, m)
        rSub = cls.getInfo(nums, m + 1, r)
        return cls.pushUp(lSub, rSub)

    @classmethod
    def pushUp(cls, l, r):
        iSum = l.iSum + r.iSum 
        lSum = max(l.lSum, l.iSum + r.lSum)
        rSum = max(r.rSum, r.iSum + l.rSum)
        mSum = max(l.mSum, r.mSum, l.rSum + r.lSum)
        return Status(lSum, rSum, mSum, iSum)

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        return Status.getInfo(nums, 0, len(nums) - 1).mSum 

File: java/0053.0_Maximum_Subarray.java
/*
DP

执行用时：1 ms, 在所有 Java 提交中击败了100.00% 的用户
内存消耗：50.7 MB, 在所有 Java 提交中击败了37.86% 的用户
通过测试用例：209 / 209
*/
class Solution {
    public int maxSubArray(int[] nums) {
        int ans = nums[0], pre = 0;
        for (int num: nums) {
            pre += num;
            ans = Math.max(ans, pre);
            if (pre < 0)
                pre = 0;
        }
        return ans;
    }
}


/*
segment tree

执行用时：8 ms, 在所有 Java 提交中击败了5.24% 的用户
内存消耗：52 MB, 在所有 Java 提交中击败了5.06% 的用户
通过测试用例：209 / 209
*/
class Solution {
    public class Status {
        public int lSum, rSum, mSum, iSum;

        public Status(int lSum, int rSum, int mSum, int iSum) {
            this.lSum = lSum;
            this.rSum = rSum;
            this.mSum = mSum;
            this.iSum = iSum;
        }
    }
    public int maxSubArray(int[] nums) {
        return getInfo(nums, 0, nums.length - 1).mSum;
    }

    public Status getInfo(int[] a, int l, int r) {
        if (l == r) {
            return new Status(a[l], a[l], a[l], a[l]);
        }
        int m = (l + r) >> 1;
        Status lSub = getInfo(a, l, m);
        Status rSub = getInfo(a, m + 1, r);
        return pushUp(lSub, rSub);
    }

    public Status pushUp(Status l, Status r) {
        int iSum = l.iSum + r.iSum;
        int lSum = Math.max(l.lSum, l.iSum + r.lSum);
        int rSum = Math.max(r.rSum, r.iSum + l.rSum);
        int mSum = Math.max(Math.max(l.mSum, r.mSum), l.rSum + r.lSum);
        return new Status(lSum, rSum, mSum, iSum);
    }
}

